### PR TITLE
Sets up a shared state for terraform 

### PR DESCRIPTION
**Pull request by goldblatt**
### Description
Addresses #1785 and forces terraform to use a shared state. I also added updates to the readme and created an initial setup file for people who are running this repo on their own machines for the first time. 

### Checklist
- [ ] Created tests which fail without the change (if possible)
- [ ] Extended the README / documentation, if necessary

### Issue(s)
Fixes #<issue_number>; Fixes #<issue_number>; Fixes #<issue_number>...

### CODE REVIEW

File: cloud/azure/initial_setup.sh [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+export ARM_ACCESS_KEY=$ACCOUNT_KEY
+
+echo "genering the backend vars file"
+touch backend_vars
+echo "resource_group_name  = \"$RESOURCE_GROUP_NAME\"" >> backend_vars
+echo "storage_account_name  = \"$STORAGE_ACCOUNT_NAME\"" >> backend_vars
+echo "container_name  = \"$CONTAINER_NAME\"" >> backend_vars
+echo "key  = \"terraform.tfstate\"" >> backend_vars
+
+
```
Comment (bion) [inline cloud/azure/initial_setup.sh:27]: suggestion (blocking): throughout file please use continuations to reduce line length below 80 chars

https://google.github.io/styleguide/shellguide.html#formatting

Comment (bion) [inline cloud/azure/initial_setup.sh:27]: suggestion (blocking): throughout file quote strings and use brackets for variable expansion

https://google.github.io/styleguide/shellguide.html#variable-expansion

File: cloud/azure/initial_setup.sh [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+CONTAINER_NAME=tfstate
+
+echo "Create resource group"
+az group create --name $RESOURCE_GROUP_NAME --location eastus
+
+echo "Create storage account"
+az storage account create --resource-group $RESOURCE_GROUP_NAME --name $STORAGE_ACCOUNT_NAME --sku Standard_LRS --encryption-services blob
+
+echo "Create blob container"
+az storage container create --name $CONTAINER_NAME --account-name $STORAGE_ACCOUNT_NAME
```
Comment (bion) [inline cloud/azure/initial_setup.sh:14]: question (if-minor): how feasible would it be to make this code idempotent?

File: cloud/azure/initial_setup.sh [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+echo "storing the account key"
+ACCOUNT_KEY=$(az storage account keys list --resource-group $RESOURCE_GROUP_NAME --account-name $STORAGE_ACCOUNT_NAME --query '[0].value' -o tsv)
+export ARM_ACCESS_KEY=$ACCOUNT_KEY
+
+echo "genering the backend vars file"
```
Comment (bion) [inline cloud/azure/initial_setup.sh:20]: suggestion (non-blocking): mention the filename/path on the local file system

File: cloud/azure/README.md [hunk 4]
```diff
@@ -1,7 +1,34 @@
...
+storage_account_name  = < account name >
+container_name  = < container name >
+key  = "terraform.tfstate"
+```
+You can get the resource group name, the storage account name, and the container name from the azure portal (or check this into a private repo). 
+
+## export your account key
+We use a shared state on an azure blob so that we don't have conflicts with our deploys. In order to get the account_key use the azure CLI. You'll need to get the resource_group_name, storage_account_name from the prior step or from someone else's backend_vars
+```
+ACCOUNT_KEY=$(az storage account keys list --resource-group $RESOURCE_GROUP_NAME --account-name $STORAGE_ACCOUNT_NAME --query '[0].value' -o tsv)
```
Comment (bion) [inline cloud/azure/README.md:17]: suggestion (non-blocking): break into multiple lines for easier readability

File: cloud/azure/README.md [hunk 5]
```diff
@@ -1,7 +1,34 @@
...
 
-## Variables
-`main.tf` uses variables via `var.var_name`. Those variable definitions and defaults live in `variables.tf`. For values that need to be private create a `.auto.tfvars` or `terraform.tfvars` file which should not be checked into a public repo (as it will often contain private information). We have added an example file to copy and fill out with specific to deployment infromation
+Backend vars should look like the following
+```
+resource_group_name  = < rg group name >
+storage_account_name  = < account name >
+container_name  = < container name >
+key  = "terraform.tfstate"
+```
```
Comment (bion) [inline cloud/azure/README.md:11]: suggestion (blocking): explicitly state that back_vars is a file that the reader needs to create in root of their project directory

File: cloud/azure/README.md [hunk 6]
```diff
@@ -1,7 +1,34 @@
...
+## running terraform with this config  
+Since we are using a shared backend you'll need to run init specifying that
+```
+$ terraform init -backend-config=backend_vars
+```
+
+# First time Setup
+In order to run for the first time we have to create the blob store to store the backend data in. 
+```
+$ chmod +x initial_setup.sh
```
Comment (bion) [inline cloud/azure/README.md:30]: suggestion (blocking): git stores the mode of each file, so this step shouldn't be necessary if the `initial_setup` script has the correct mode when checked in

File: cloud/azure/initial_setup.sh [hunk 7]
```diff
@@ -0,0 +1,27 @@
+#!/bin/bash
```
Comment (bion) [inline cloud/azure/initial_setup.sh:1]: suggestion (blocking): remove `.sh` filename extension as per styleguide

File: cloud/azure/README.md [hunk 8]
```diff
@@ -1,7 +1,34 @@
...
+export ARM_ACCESS_KEY=$ACCOUNT_KEY
+```
+
+## running terraform with this config  
+Since we are using a shared backend you'll need to run init specifying that
+```
+$ terraform init -backend-config=backend_vars
+```
+
+# First time Setup
```
Comment (bion) [inline cloud/azure/README.md:27]:
```suggestion
# First time setup
```

File: cloud/azure/README.md [hunk 9]
```diff
@@ -1,7 +1,34 @@
-# Azure Terraform 
+# Azure Terraform Setup
```
Comment (bion) [inline cloud/azure/README.md:1]: suggestion (blocking): are there any values generated during these steps that should be considered secrets? if so please specify

File: cloud/azure/README.md [hunk 10]
```diff
@@ -1,7 +1,34 @@
...
+```
+resource_group_name  = < rg group name >
+storage_account_name  = < account name >
+container_name  = < container name >
+key  = "terraform.tfstate"
+```
+You can get the resource group name, the storage account name, and the container name from the azure portal (or check this into a private repo). 
+
+## export your account key
+We use a shared state on an azure blob so that we don't have conflicts with our deploys. In order to get the account_key use the azure CLI. You'll need to get the resource_group_name, storage_account_name from the prior step or from someone else's backend_vars
```
Comment (shubha-rajan) [inline cloud/azure/README.md:15]: suggestion(blocking): can you add a line explaining what this account key value is used for?
```suggestion
In order to get the `account_key` value, use the azure CLI. You'll need to get the resource_group_name, storage_account_name from the prior step.
```

File: cloud/azure/main.tf [hunk 11]
```diff
@@ -5,17 +5,20 @@ terraform {
...
-
+  backend "azurerm" {}
   required_version = ">= 0.14.9"
 }
 
 provider "azurerm" {
   features {}
 }
 
+resource "random_pet" "server" {}
```
Comment (bion) [inline cloud/azure/main.tf:18]: 🐮

File: cloud/azure/README.md [hunk 12]
```diff
@@ -1,7 +1,34 @@
-# Azure Terraform 
+# Azure Terraform Setup
```
Comment (shubha-rajan) [inline cloud/azure/README.md:2]: thought: I think this line will work better as an intro at the top of the file than in the subsection it was in.
```suggestion
We use a remote backend to store state information to avoid conflicts when multiple collaborators are updating resources managed by Terraform. Follow the steps outlined in this doc to enable collaborators to access the remote backend and shared state on their machines.
```

File: cloud/azure/README.md [hunk 13]
```diff
@@ -1,7 +1,34 @@
...
+container_name  = < container name >
+key  = "terraform.tfstate"
+```
+You can get the resource group name, the storage account name, and the container name from the azure portal (or check this into a private repo). 
+
+## export your account key
+We use a shared state on an azure blob so that we don't have conflicts with our deploys. In order to get the account_key use the azure CLI. You'll need to get the resource_group_name, storage_account_name from the prior step or from someone else's backend_vars
+```
+ACCOUNT_KEY=$(az storage account keys list --resource-group $RESOURCE_GROUP_NAME --account-name $STORAGE_ACCOUNT_NAME --query '[0].value' -o tsv)
+export ARM_ACCESS_KEY=$ACCOUNT_KEY
```
Comment (shubha-rajan) [inline cloud/azure/README.md:18]: question(blocking): will running these lines be necessary each time a collaborator uses terraform? or just for the init?

File: cloud/azure/main.tf [hunk 14]
```diff
@@ -60,7 +63,7 @@ resource "azurerm_app_service_plan" "plan" {
 }
 
 resource "azurerm_app_service" "civiform_app" {
-  name                = var.application_name
+  name                = "${var.application_name}-${random_pet.server.id}"
```
Comment (shubha-rajan) [inline cloud/azure/main.tf:66]: question(blocking): is this to ensure app and db names are unique? and if so, should it be done in another PR since it doesn't affect the backend config?

Comment (shubha-rajan) [reply cloud/azure/main.tf:66]: also random_pet is the best terraform random resource. numbers are so boring!

File: cloud/azure/README.md (hunk 9 above)
Comment (goldblatt) [reply cloud/azure/README.md:1]: no there aren't actually it's just specific to your setup

Comment (bion) [reply cloud/azure/README.md:1]: ah nice, thanks!

File: cloud/azure/main.tf (hunk 14 above)
Comment (goldblatt) [reply cloud/azure/main.tf:66]: defffinitely could have been in another PR as they were tangentially related (couldn't really test this PR without fixing that, but might be a lil bit of a pain to split up now)

File: cloud/azure/README.md (hunk 6 above)
Comment (goldblatt) [reply cloud/azure/README.md:30]: huh TIL

File: cloud/azure/README.md (hunk 13 above)
Comment (goldblatt) [reply cloud/azure/README.md:18]: yess you need the arm_access_key for all the tf commands

File: cloud/azure/README.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (goldblatt) [reply cloud/azure/README.md]: alright swapped the order around and added some extra specificity

File: cloud/azure/README.md (hunk 13 above)
Comment (goldblatt) [reply cloud/azure/README.md:18]: I added a mini troubleshooting message

File: cloud/azure/README.md [hunk 15]
```diff
@@ -1,7 +1,34 @@
...
-## Variables
-`main.tf` uses variables via `var.var_name`. Those variable definitions and defaults live in `variables.tf`. For values that need to be private create a `.auto.tfvars` or `terraform.tfvars` file which should not be checked into a public repo (as it will often contain private information). We have added an example file to copy and fill out with specific to deployment infromation
+Backend vars should look like the following
+```
+resource_group_name  = < rg group name >
+storage_account_name  = < account name >
+container_name  = < container name >
+key  = "terraform.tfstate"
+```
+You can get the resource group name, the storage account name, and the container name from the azure portal (or check this into a private repo).
```
[the remark this answers is not in the document]
Comment (goldblatt) [reply cloud/azure/README.md]: added one

File: cloud/azure/initial_setup.sh (hunk 2 above)
Comment (goldblatt) [reply cloud/azure/initial_setup.sh:14]: yeahhh was thinking about this itsss maybe a lil tricky due to the needs to be globally unique aspect. basically i don't think it's totally worth the effort but i added some extra docs. maybe could use something like git branch or date or some other like less random but still unique thing here instead which wouldn't be idempotent but would be maybe less spooky

File: cloud/azure/main.tf (hunk 14 above)
Comment (shubha-rajan) [reply cloud/azure/main.tf:66]: I'm fine with leaving it in.

File: cloud/azure/initial_setup.sh (hunk 2 above)
Comment (goldblatt) [reply cloud/azure/initial_setup.sh]: okay yes shubha and i chatted about this in our sync and am gonna use the like if it exists then do x

Comment (bion) [reply cloud/azure/initial_setup.sh]: great!

File: cloud/azure/README.md [hunk 16]
```diff
@@ -1,7 +1,85 @@
...
+
+Note that it takes azure a few minutes for you to be able to access the blob 
+storage and that you'll need to share the backend_vars with other people 
+developing on your project. You can find the info on the azure portal as well. 
+
+You can run this command again if you delete your backend_vars or need to
+re-export your account key (although you can also re-export your account key
+via the commands below), note that it does a pretty fuzzy regex match for
+the storage account name so if you are doing something special you might have
+to wright your own script to do that!
```
Comment (bion) [inline cloud/azure/README.md:30]:
```suggestion
to write your own script to do that!
```

File: cloud/azure/README.md [hunk 17]
```diff
@@ -1,7 +1,85 @@
...
+
+## Export your account key
+In order to use terraform with the azure backend you'll need to get the 
+`account_key` set up properly. The backend_vars file expects it to be provided 
+via ARM_ACCESS_KEY so you'll need to use the Azure CLI to provide it.
```
Comment (bion) [inline cloud/azure/README.md:35]:
```suggestion
via the `ARM_ACCESS_KEY` environment variable so you'll need to use the Azure CLI to provide it. 
```

File: cloud/azure/README.md [hunk 18]
```diff
@@ -1,7 +1,85 @@
...
+`account_key` set up properly. The backend_vars file expects it to be provided 
+via ARM_ACCESS_KEY so you'll need to use the Azure CLI to provide it. 
+Get the `resource_group_name` and `storage_account_name` from the backend_vars
+file or via the azure cli/portal (details on how to do this in the 
+troubleshooting section below)
+```
+ACCOUNT_KEY=$(az storage account keys list --resource-group \
+$RESOURCE_GROUP_NAME --account-name $STORAGE_ACCOUNT_NAME --query \
+'[0].value' -o tsv)
+export ARM_ACCESS_KEY=$ACCOUNT_KEY
```
Comment (bion) [inline cloud/azure/README.md:43]:
```suggestion
ACCOUNT_KEY=$(az storage account keys list \
  --resource-group $RESOURCE_GROUP_NAME \
  --account-name $STORAGE_ACCOUNT_NAME \
  --query '[0].value' \
  -o tsv)
export ARM_ACCESS_KEY=$ACCOUNT_KEY
```

File: cloud/azure/README.md [hunk 19]
```diff
@@ -1,7 +1,85 @@
...
+```
+
+## Running terraform with this config  
+Since we are using a shared backend you'll need to run init specifying that
+```
+$ terraform init -backend-config=backend_vars
```
Comment (bion) [inline cloud/azure/README.md:49]: question (blocking): please mention which pwd the reader should use when running these commands

File: cloud/azure/README.md [hunk 20]
```diff
@@ -1,7 +1,85 @@
...
+# Troubleshooting
+
+## Access Issues
+If you are running into access issues, you do need to have the ARM_ACCESS_KEY 
+exported to run terraform commands. So try re running
+```
+ACCOUNT_KEY=$(az storage account keys list --resource-group \
+$RESOURCE_GROUP_NAME --account-name $STORAGE_ACCOUNT_NAME --query \
+'[0].value' -o tsv)
+export ARM_ACCESS_KEY=$ACCOUNT_KEY
```
Comment (bion) [inline cloud/azure/README.md:61]:
```suggestion
ACCOUNT_KEY=$(az storage account keys list \
  --resource-group $RESOURCE_GROUP_NAME \
  --account-name $STORAGE_ACCOUNT_NAME \
  --query '[0].value' \
  -o tsv)
export ARM_ACCESS_KEY=$ACCOUNT_KEY
```

File: cloud/azure/README.md [hunk 21]
```diff
@@ -1,7 +1,85 @@
...
+
+## backend_vars file structure
+
+This is used by terraform to specify the backend we wil use to store tfsate 
+and generally should look like this
+```
+resource_group_name  = tfstate
+storage_account_name  = < account name >
+container_name  = tfstate
+key  = "terraform.tfstate"
```
Comment (bion) [inline cloud/azure/README.md:72]:
```suggestion
resource_group_name  = tfstate
storage_account_name = < account name >
container_name       = tfstate
key                  = "terraform.tfstate"
```

File: cloud/azure/setup [hunk 22]
```diff
@@ -0,0 +1,56 @@
...
+    echo "Creating the container"
+    az storage container create --name ${CONTAINER_NAME} \
+    --account-name ${STORAGE_ACCOUNT_NAME}
+fi
+
+echo "storing the account key"
+ACCOUNT_KEY="$(az storage account keys list --resource-group \
+${RESOURCE_GROUP_NAME} --account-name ${STORAGE_ACCOUNT_NAME} \
+--query '[0].value' -o tsv)"
+export ARM_ACCESS_KEY="${ACCOUNT_KEY}"
```
Comment (bion) [inline cloud/azure/setup:48]: nit: please indent continuations and use one flag/value pair per line

File: cloud/azure/setup [hunk 23]
```diff
@@ -0,0 +1,56 @@
...
+    STORAGE_ACCOUNT_NAME=${BASH_REMATCH[1]}
+else
+    echo "Create storage account"
+    az storage account create --resource-group ${RESOURCE_GROUP_NAME} \
+    --name ${STORAGE_ACCOUNT_NAME} --sku Standard_LRS --encryption-services blob
+fi
+
+echo "Create blob container if it doesn't exist"
+CONTAINER_EXIST=$(az storage container exists --name ${CONTAINER_NAME} \
+--account-name ${STORAGE_ACCOUNT_NAME})
```
Comment (bion) [inline cloud/azure/setup:34]: nit: please indent continuations and use one flag/value pair per line

File: cloud/azure/setup [hunk 24]
```diff
@@ -0,0 +1,56 @@
...
+${RESOURCE_GROUP_NAME})
+
+STORAGE_ACCOUNT_PATTERN="\"name\"\: \"(tfstate[0-9]+)\""
+if [[ $CURRENT_ACCOUNTS =~ $STORAGE_ACCOUNT_PATTERN ]]; then
+    echo "You have a storage account already, using the name ${BASH_REMATCH[1]}"
+    STORAGE_ACCOUNT_NAME=${BASH_REMATCH[1]}
+else
+    echo "Create storage account"
+    az storage account create --resource-group ${RESOURCE_GROUP_NAME} \
+    --name ${STORAGE_ACCOUNT_NAME} --sku Standard_LRS --encryption-services blob
```
Comment (bion) [inline cloud/azure/setup:29]: nit: please indent continuations and use one flag/value pair per line

File: cloud/azure/setup [hunk 25]
```diff
@@ -0,0 +1,56 @@
...
+ACCOUNT_KEY="$(az storage account keys list --resource-group \
+${RESOURCE_GROUP_NAME} --account-name ${STORAGE_ACCOUNT_NAME} \
+--query '[0].value' -o tsv)"
+export ARM_ACCESS_KEY="${ACCOUNT_KEY}"
+
+echo "generating the backend_vars file"
+echo "resource_group_name  = \"${RESOURCE_GROUP_NAME}\"" > backend_vars
+echo "storage_account_name  = \"${STORAGE_ACCOUNT_NAME}\"" >> backend_vars
+echo "container_name  = \"${CONTAINER_NAME}\"" >> backend_vars
+echo "key  = \"terraform.tfstate\"" >> backend_vars
```
Comment (bion) [inline cloud/azure/setup:54]: nit: align `backend_vars` on `=` to fit tf style

**Review by bion — APPROVED** [review]
